openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+/// narrow-band level sets (of width ~2x3). Rather it is an efficient
+/// technique when distance values must be computed far from known
+/// distance values (typically close to an interface).
+///
+/// @warning This class is for experts only (the sequence in with the different
+///          methods are called is important). Instead call one of the free-standing
+///          functions listed above!
+template<typename GridT>
+class FastSweeping
+{
+    using ValueT = typename GridT::ValueType;
+    using TreeT  = typename GridT::TreeType;
+    using AccT   = tree::ValueAccessor<TreeT, false>;//don't register accessors
+
+    // This class can only be templated on a grid with floating-point values!
+    static_assert(std::is_floating_point<ValueT>::value,
+                  "FastSweeping requires a grid with floating-point values");
+public:
+
+    /// @brief Constructor
+    FastSweeping();
+
+     /// @brief Destructor.
+    ~FastSweeping() { this->clear(); }
+
+    /// @brief Disallow copy construction.
+    FastSweeping(const FastSweeping&) = delete;
+
+    /// @brief Disallow copy assignment.
+    FastSweeping& operator=(const FastSweeping&) = delete;
+
+    /// @brief Returns a shared pointer to the signed distance field computed
+    ///        by this class.
+    /// @warning This shared pointer might point to NULL if the grid has not been
+    ///          initialize (by one of the init methods) or computed (by the sweep
+    ///          method).
+    typename GridT::Ptr sdfGrid() { return mGrid1; }
+    typename GridT::Ptr extGrid() { return mGrid2; }
+
+    void initSdf(const GridT &fogGrid, ValueT isoValue, bool isInputSdf);
+
+    template <typename OpT>
+    void initExt(const GridT &fogGrid, const OpT &op, ValueT isoValue, bool isInputSdf);
+
+    // use tools::NN_FACE_EDGE for improved dilation
+    void initDilate(const GridT &sdfGrid, int dilation, NearestNeighbors nn = NN_FACE);
+
+    template<typename MaskTreeT>
+    void initMask(const GridT &sdfGrid, const Grid<MaskTreeT> &mask, bool ignoreActiveTiles = false);
+
+    /// @brief Perform @a nIter iterations of the fast sweeping algorithm.
+    void sweep(int nIter = 1, bool finalize = true);
+
+    void clear();// { this->init(nullptr); }
+
+    /// @brief Return the number of voxels that will be solved for.
+    size_t voxelCount() const { return mPagedArray.size(); }
+
+    /// @brief Return the number of voxels that defined the boundary condition.
+    size_t boundaryCount() const { return mBoundaryCount; }
+private:
+
+    // Private classes to initialize the grid and construct
+    // mPagedArray with voxel coordinates.
+    template<typename>
+    struct MaskKernel;//   initialization to extand a SDF into a mask
+    template<typename>
+    struct InitExt;
+    struct InitSdf;
+    struct DilateKernel;// initialization to dilate a SDF
+    struct MinMaxKernel;
+    struct SweepingKernel;// Private class to perform the actual concurrent sparse fast sweeping
```

Comment:
This is a broader design question and I'm not asking or implying we should meke any changes to the internals now - but what's the benefit of these private sub structs? I'd instead propose these were inline lambdas within the forward api methods i.e. `initDilate` doesn't construct a `DilateKernel` but instead implements `DilateKernel::run` itself

Reply:
they are used multiple times and each of them have multiple member methods